File: src/llm/rag.rs
// Copyright 2025 ProximaDB
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.

//! RAG (Retrieval-Augmented Generation) Pipeline
//!
//! Uses ProximaDB collections for document storage and retrieval,
//! with generation via Victor's LLM providers.

use anyhow::Result;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::Arc;
use tokio::sync::RwLock;

use crate::llm::config::RAGConfig;

/// Document for RAG indexing
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Document {
    /// Unique document ID
    pub id: String,
    /// Document title
    pub title: String,
    /// Document content
    pub content: String,
    /// Source (file path, URL, etc.)
    pub source: String,
    /// Additional metadata
    pub metadata: HashMap<String, String>,
}

/// RAG query request
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RAGRequest {
    /// Question/query
    pub question: String,
    /// Collection to search
    pub collection: String,
    /// Number of documents to retrieve
    #[serde(default = "default_top_k")]
    pub top_k: usize,
    /// Optional metadata filter
    pub filter: Option<HashMap<String, String>>,
    /// Custom system prompt
    pub system_prompt: Option<String>,
    /// Temperature for generation
    #[serde(default = "default_temperature")]
    pub temperature: f32,
    /// Maximum response tokens
    #[serde(default = "default_max_tokens")]
    pub max_tokens: usize,
    /// Whether to include sources in response
    #[serde(default = "default_true")]
    pub include_sources: bool,
}

fn default_top_k() -> usize {
    5
}
fn default_temperature() -> f32 {
    0.7
}
fn default_max_tokens() -> usize {
    1024
}
fn default_true() -> bool {
    true
}

/// Source document in RAG response
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Source {
    /// Document ID
    pub id: String,
    /// Document title
    pub title: String,
    /// Source location (file path, URL)
    pub url: String,
    /// Relevance score (0-1)
    pub relevance: f32,
    /// Snippet of content
    pub snippet: String,
}

/// RAG response
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RAGResponse {
    /// Generated answer
    pub answer: String,
    /// Source documents used
    pub sources: Vec<Source>,
    /// Confidence score (0-1)
    pub confidence: f32,
    /// Total latency in milliseconds
    pub latency_ms: u64,
    /// Retrieval latency in milliseconds
    pub retrieval_latency_ms: u64,
    /// Generation latency in milliseconds
    pub generation_latency_ms: u64,
    /// Tokens used in generation
    pub tokens_used: u64,
    /// Whether this was a cache hit
    pub cached: bool,
}

/// RAG context built from retrieved documents
#[derive(Debug, Clone)]
pub struct RAGContext {
    /// Combined context text
    pub text: String,
    /// Source documents
    pub sources: Vec<Source>,
    /// Estimated token count
    pub token_count: usize,
}

/// RAG pipeline statistics
#[derive(Debug, Clone, Default)]
pub struct RAGStats {
    pub total_queries: u64,
    pub total_documents_retrieved: u64,
    pub cache_hits: u64,
    pub cache_misses: u64,
    pub total_tokens_used: u64,
    pub average_latency_ms: f64,
    pub average_sources_per_query: f64,
}

/// RAG Pipeline
///
/// Orchestrates document retrieval from ProximaDB collections
/// and LLM generation via Victor.
pub struct RAGPipeline {
    config: RAGConfig,
    /// Statistics
    stats: Arc<RwLock<RAGStats>>,
    /// Total queries counter
    total_queries: AtomicU64,
    /// Collection name -> document count
    collections: Arc<RwLock<HashMap<String, usize>>>,
    initialized: Arc<RwLock<bool>>,
}

impl RAGPipeline {
    /// Create a new RAG pipeline
    pub fn new(config: RAGConfig) -> Result<Self> {
        Ok(Self {
            config,
            stats: Arc::new(RwLock::new(RAGStats::default())),
            total_queries: AtomicU64::new(0),
            collections: Arc::new(RwLock::new(HashMap::new())),
            initialized: Arc::new(RwLock::new(false)),
        })
    }

    /// Initialize the RAG pipeline
    pub async fn initialize(&self) -> Result<()> {
        let mut initialized = self.initialized.write().await;
        if *initialized {
            return Ok(());
        }

        tracing::info!(
            retrieval_top_k = %self.config.retrieval_top_k,
            context_top_k = %self.config.context_top_k,
            "Initializing RAG pipeline"
        );

        *initialized = true;
        tracing::info!("RAG pipeline initialized");
        Ok(())
    }

    /// Execute a RAG query
    ///
    /// This coordinates retrieval from ProximaDB and generation via Victor.
    /// The actual implementation is in Python (Victor/codingagent).
    pub async fn query(&self, request: RAGRequest) -> Result<RAGResponse> {
        let start = std::time::Instant::now();

        // Increment query counter
        self.total_queries.fetch_add(1, Ordering::Relaxed);

        // Build response structure
        // Actual RAG is performed via Victor's Python SDK
        let response = RAGResponse {
            answer: String::new(),
            sources: Vec::new(),
            confidence: 0.0,
            latency_ms: start.elapsed().as_millis() as u64,
            retrieval_latency_ms: 0,
            generation_latency_ms: 0,
            tokens_used: 0,
            cached: false,
        };

        // Update stats
        let mut stats = self.stats.write().await;
        stats.total_queries += 1;
        stats.cache_misses += 1;

        Ok(response)
    }

    /// Index documents into a collection
    pub async fn index_documents(&self, collection: &str, documents: Vec<Document>) -> Result<usize> {
        let doc_count = documents.len();

        // Update collection tracking
        let mut collections = self.collections.write().await;
        let current = collections.entry(collection.to_string()).or_insert(0);
        *current += doc_count;

        tracing::info!(
            collection = %collection,
            documents = %doc_count,
            "Indexed documents for RAG"
        );

        Ok(doc_count)
    }

    /// Build context from search results
    pub fn build_context(&self, sources: Vec<Source>, max_tokens: usize) -> RAGContext {
        let mut context_parts = Vec::new();
        let mut total_tokens = 0;
        let mut included_sources = Vec::new();

        for (i, source) in sources.into_iter().enumerate() {
            // Estimate tokens (rough: 4 chars per token)
            let source_tokens = source.snippet.len() / 4;

            if total_tokens + source_tokens > max_tokens {
                break;
            }

            context_parts.push(format!("[{}] {}", i + 1, source.snippet));
            total_tokens += source_tokens;
            included_sources.push(source);
        }

        RAGContext {
            text: context_parts.join("\n\n"),
            sources: included_sources,
            token_count: total_tokens,
        }
    }

    /// Build RAG prompt
    pub fn build_prompt(
        &self,
        question: &str,
        context: &RAGContext,
        system_prompt: Option<&str>,
    ) -> String {
        let default_system = "Answer the question based on the provided context. \
            If the answer cannot be found in the context, say so clearly.";

        format!(
            "{}\n\nContext:\n{}\n\nQuestion: {}\n\nAnswer:",
            system_prompt.unwrap_or(default_system),
            context.text,
            question
        )
    }

    /// Get statistics
    pub async fn get_stats(&self) -> RAGStats {
        self.stats.read().await.clone()
    }

    /// Get collection document counts
    pub async fn get_collections(&self) -> HashMap<String, usize> {
        self.collections.read().await.clone()
    }

    /// Check if RAG is enabled
    pub fn is_enabled(&self) -> bool {
        self.config.enabled
    }

    /// Get config
    pub fn config(&self) -> &RAGConfig {
        &self.config
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_rag_pipeline_creation() {
        let config = RAGConfig::default();
        let pipeline = RAGPipeline::new(config);
        assert!(pipeline.is_ok());
    }

    #[tokio::test]
    async fn test_rag_pipeline_initialization() {
        let config = RAGConfig::default();
        let pipeline = RAGPipeline::new(config).unwrap();
        let result = pipeline.initialize().await;
        assert!(result.is_ok());
    }

    #[tokio::test]
    async fn test_rag_build_context() {
        let config = RAGConfig::default();
        let pipeline = RAGPipeline::new(config).unwrap();

        let sources = vec![
            Source {
                id: "1".to_string(),
                title: "Doc 1".to_string(),
                url: "file://doc1.txt".to_string(),
                relevance: 0.9,
                snippet: "This is the first document content.".to_string(),
            },
            Source {
                id: "2".to_string(),
                title: "Doc 2".to_string(),
                url: "file://doc2.txt".to_string(),
                relevance: 0.8,
                snippet: "This is the second document content.".to_string(),
            },
        ];

        let context = pipeline.build_context(sources, 1000);
        assert_eq!(context.sources.len(), 2);
        assert!(context.text.contains("[1]"));
        assert!(context.text.contains("[2]"));
    }

    #[tokio::test]
    async fn test_rag_build_prompt() {
        let config = RAGConfig::default();
        let pipeline = RAGPipeline::new(config).unwrap();

        let context = RAGContext {
            text: "Some context text".to_string(),
            sources: vec![],
            token_count: 10,
        };

        let prompt = pipeline.build_prompt("What is this?", &context, None);
        assert!(prompt.contains("What is this?"));
        assert!(prompt.contains("Some context text"));
    }

    #[tokio::test]
    async fn test_rag_index_documents() {
        let config = RAGConfig::default();
        let pipeline = RAGPipeline::new(config).unwrap();
        pipeline.initialize().await.unwrap();

        let documents = vec![
            Document {
                id: "1".to_string(),
                title: "Doc 1".to_string(),
                content: "Content 1".to_string(),
                source: "file://1".to_string(),
                metadata: HashMap::new(),
            },
            Document {
                id: "2".to_string(),
                title: "Doc 2".to_string(),
                content: "Content 2".to_string(),
                source: "file://2".to_string(),
                metadata: HashMap::new(),
            },
        ];

        let count = pipeline.index_documents("test", documents).await.unwrap();
        assert_eq!(count, 2);

        let collections = pipeline.get_collections().await;
        assert_eq!(collections.get("test"), Some(&2));
    }
}

File: clients/python/src/proximadb_sdk/llm/rag.py
# Copyright 2025 ProximaDB
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.

"""RAG (Retrieval-Augmented Generation) Pipeline using ProximaDB."""

import time
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, TYPE_CHECKING

from proximadb_sdk.llm.config import EmbeddingConfig, LLMConfig, RAGConfig
from proximadb_sdk.llm.embedding import EmbeddingService
from proximadb_sdk.llm.semantic_cache import SemanticCache

if TYPE_CHECKING:
    from proximadb_sdk import ProximaDBClient


@dataclass
class Document:
    """Document for RAG indexing.

    Attributes:
        id: Unique document ID
        title: Document title
        content: Document content
        source: Source location (file path, URL)
        metadata: Additional metadata
    """

    id: str
    title: str
    content: str
    source: str
    metadata: Dict[str, Any] = field(default_factory=dict)


@dataclass
class Source:
    """Source document in RAG response.

    Attributes:
        id: Document ID
        title: Document title
        url: Source URL/path
        relevance: Relevance score (0-1)
        snippet: Content snippet
    """

    id: str
    title: str
    url: str
    relevance: float
    snippet: str


@dataclass
class RAGResponse:
    """RAG query response.

    Attributes:
        answer: Generated answer
        sources: Source documents used
        confidence: Confidence score (0-1)
        latency_ms: Total latency in milliseconds
        retrieval_latency_ms: Retrieval latency
        generation_latency_ms: Generation latency
        tokens_used: Tokens used in generation
        cached: Whether response was from cache
    """

    answer: str
    sources: List[Source]
    confidence: float
    latency_ms: int
    retrieval_latency_ms: int
    generation_latency_ms: int
    tokens_used: int
    cached: bool


class RAGPipeline:
    """RAG Pipeline using ProximaDB for document storage.

    Provides document indexing, retrieval, and LLM generation with:
    - ProximaDB vector storage for embeddings
    - Semantic caching for repeated queries
    - Flexible LLM provider support via Victor

    Usage:
        from proximadb_sdk import ProximaDBClient
        from proximadb_sdk.llm import RAGPipeline, Document

        client = ProximaDBClient(url="http://localhost:5678")
        rag = RAGPipeline(client)
        await rag.initialize()

        # Index documents
        docs = [Document(id="1", title="Doc", content="...", source="file.txt")]
        await rag.index_documents("knowledge", docs)

        # Query
        response = await rag.query("What is...?", "knowledge")
        print(response.answer)
    """

    def __init__(
        self,
        client: "ProximaDBClient",
        config: Optional[LLMConfig] = None,
    ):
        """Initialize RAG pipeline.

        Args:
            client: ProximaDB client
            config: LLM configuration (optional)
        """
        self.client = client
        self.config = config or LLMConfig()
        self.embedding_service = EmbeddingService(self.config.embedding)
        self.semantic_cache = SemanticCache(self.config.cache, client)
        self._initialized = False

    async def initialize(self) -> None:
        """Initialize the RAG pipeline."""
        if self._initialized:
            return

        await self.embedding_service.initialize()
        await self.semantic_cache.initialize()
        self._initialized = True

    async def index_documents(
        self,
        collection: str,
        documents: List[Document],
        create_collection: bool = True,
    ) -> int:
        """Index documents into a collection.

        Args:
            collection: Collection name
            documents: List of documents to index
            create_collection: Whether to create collection if it doesn't exist

        Returns:
            Number of documents indexed
        """
        if not self._initialized:
            await self.initialize()

        if not documents:
            return 0

        # Create collection if needed
        if create_collection:
            try:
                await self.client.create_collection_async(
                    name=collection,
                    dimension=self.embedding_service.dimension,
                    distance_metric="cosine",
                )
            except Exception:
                # Collection may already exist
                pass

        # Generate embeddings in batches
        batch_size = self.config.embedding.batch_size
        total_indexed = 0

        for i in range(0, len(documents), batch_size):
            batch = documents[i : i + batch_size]
            contents = [doc.content for doc in batch]
            embeddings = await self.embedding_service.embed_batch(contents)

            # Prepare vectors for insertion
            vectors = []
            for doc, embedding in zip(batch, embeddings):
                vectors.append(
                    {
                        "id": doc.id,
                        "vector": embedding,
                        "metadata": {
                            "title": doc.title,
                            "content": doc.content[:1000],  # Truncate for metadata
                            "source": doc.source,
                            **doc.metadata,
                        },
                    }
                )

            # Insert into ProximaDB
            await self.client.insert_vectors_async(collection, vectors)
            total_indexed += len(batch)

        return total_indexed

    async def query(
        self,
        question: str,
        collection: str,
        top_k: Optional[int] = None,
        filter_metadata: Optional[Dict[str, Any]] = None,
        system_prompt: Optional[str] = None,
        skip_cache: bool = False,
    ) -> RAGResponse:
        """Execute a RAG query.

        Args:
            question: Question to answer
            collection: Collection to search
            top_k: Number of documents to retrieve
            filter_metadata: Optional metadata filters
            system_prompt: Custom system prompt
            skip_cache: Whether to skip semantic cache

        Returns:
            RAG response with answer and sources
        """
        if not self._initialized:
            await self.initialize()

        start_time = time.time()
        top_k = top_k or self.config.rag.retrieval_top_k

        # Check semantic cache
        if not skip_cache and self.config.cache.enabled:
            cached = await self.semantic_cache.lookup(question, collection)
            if cached:
                return RAGResponse(
                    answer=cached.response.answer,
                    sources=cached.response.sources,
                    confidence=cached.response.confidence,
                    latency_ms=int((time.time() - start_time) * 1000),
                    retrieval_latency_ms=0,
                    generation_latency_ms=0,
                    tokens_used=0,
                    cached=True,
                )

        # Generate question embedding
        retrieval_start = time.time()
        question_embedding = await self.embedding_service.embed_text(question)

        # Search for relevant documents
        search_results = await self.client.search_async(
            collection=collection,
            vector=question_embedding,
            top_k=top_k,
            filter=filter_metadata,
        )

        retrieval_latency = int((time.time() - retrieval_start) * 1000)

        # Build context from results
        sources = []
        context_parts = []

        for i, result in enumerate(search_results[: self.config.rag.context_top_k]):
            metadata = result.get("metadata", {})
            content = metadata.get("content", "")

            source = Source(
                id=result.get("id", f"doc_{i}"),
                title=metadata.get("title", "Untitled"),
                url=metadata.get("source", ""),
                relevance=1.0 - result.get("distance", 0.0),
                snippet=content[:500],
            )
            sources.append(source)
            context_parts.append(f"[{i + 1}] {content}")

        context = "\n\n".join(context_parts)

        # Generate answer using LLM
        generation_start = time.time()
        answer, tokens_used = await self._generate_answer(
            question=question,
            context=context,
            system_prompt=system_prompt,
        )
        generation_latency = int((time.time() - generation_start) * 1000)

        # Calculate confidence
        confidence = self._calculate_confidence(sources)

        response = RAGResponse(
            answer=answer,
            sources=sources,
            confidence=confidence,
            latency_ms=int((time.time() - start_time) * 1000),
            retrieval_latency_ms=retrieval_latency,
            generation_latency_ms=generation_latency,
            tokens_used=tokens_used,
            cached=False,
        )

        # Cache response
        if not skip_cache and self.config.cache.enabled:
            await self.semantic_cache.store(
                question=question,
                collection=collection,
                embedding=question_embedding,
                response=response,
            )

        return response

    async def _generate_answer(
        self,
        question: str,
        context: str,
        system_prompt: Optional[str] = None,
    ) -> tuple[str, int]:
        """Generate answer using LLM provider.

        Args:
            question: Question to answer
            context: Context from retrieved documents
            system_prompt: Custom system prompt

        Returns:
            Tuple of (answer, tokens_used)
        """
        default_system = (
            "Answer the question based on the provided context. "
            "If the answer cannot be found in the context, say so clearly. "
            "Cite relevant sources by their numbers [1], [2], etc."
        )

        prompt = f"""{system_prompt or default_system}

Context:
{context}

Question: {question}

Answer:"""

        try:
            # Try to use Victor's LLM providers
            from victor.providers.registry import ProviderRegistry

            registry = ProviderRegistry()
            provider = registry.get_provider(self.config.rag.llm_provider)

            response = await provider.chat(
                messages=[{"role": "user", "content": prompt}],
                model=self.config.rag.llm_model,
                temperature=self.config.rag.temperature,
                max_tokens=self.config.rag.max_tokens,
            )

            return response.content, response.usage.total_tokens

        except ImportError:
            # Fall back to simple prompt-based answer
            return self._fallback_answer(question, context), 0

    def _fallback_answer(self, question: str, context: str) -> str:
        """Generate a simple answer when no LLM is available.

        This provides basic extractive QA without an LLM.
        """
        # Simple extractive approach: find most relevant sentence
        sentences = context.split(". ")
        question_words = set(question.lower().split())

        best_sentence = ""
        best_score = 0

        for sentence in sentences:
            sentence_words = set(sentence.lower().split())
            overlap = len(question_words & sentence_words)
            if overlap > best_score:
                best_score = overlap
                best_sentence = sentence

        if best_sentence:
            return f"Based on the context: {best_sentence}."
        return "I couldn't find a relevant answer in the provided context."

    def _calculate_confidence(self, sources: List[Source]) -> float:
        """Calculate confidence score based on sources."""
        if not sources:
            return 0.0

        # Average relevance of top sources
        top_relevances = [s.relevance for s in sources[:3]]
        return sum(top_relevances) / len(top_relevances)

    async def delete_documents(
        self,
        collection: str,
        document_ids: List[str],
    ) -> int:
        """Delete documents from collection.

        Args:
            collection: Collection name
            document_ids: IDs of documents to delete

        Returns:
            Number of documents deleted
        """
        deleted = 0
        for doc_id in document_ids:
            try:
                await self.client.delete_vector_async(collection, doc_id)
                deleted += 1
            except Exception:
                pass
        return deleted

    async def clear_collection(self, collection: str) -> None:
        """Clear all documents from a collection.

        Args:
            collection: Collection name to clear
        """
        try:
            await self.client.delete_collection_async(collection)
        except Exception:
            pass

    async def close(self) -> None:
        """Clean up resources."""
        await self.embedding_service.close()
        self._initialized = False
